scanpipe/pipes/d2d.py

```diff
@@ -524,3 +526,110 @@ def match_purldb(project, extensions, logger=None):
+    # side we will need to supply another layer of logic to selectively get those
+    # specific 7-8 file that might relate to a particular file from `from` side.
+
+    if logger:
+        logger(
+            f"Mapping {resource_count:,d} from/ resources using javascript map "
+            f"against to/ codebase"
+        )
+
+    to_resources_index = pathmap.build_index(
+        to_resources.values_list("id", "path"), with_subpaths=True
+    )
+
+    resource_iterator = from_resources.iterator(chunk_size=2000)
+    last_percent = 0
+    start_time = timer()
+    for resource_index, from_resource in enumerate(resource_iterator):
+        last_percent = pipes.log_progress(
+            logger,
+            resource_index,
+            resource_count,
+            last_percent,
+            increment_percent=10,
+            start_time=start_time,
+        )
+        _map_javascript_resource(from_resource, to_resources_index, to_resources)
+
+
+def _map_javascript_resource(from_resource, to_resources_index, to_resources):
+    path = Path(from_resource.path.lstrip("/"))
+
+    basename, from_extension = js.get_basename_and_extension(path.name)
+    path_parts = (path.parent / basename).parts
+    path_parts_len = len(path_parts)
+
+    base_path = path.parent / basename
+
+    candidates = PROSPECTIVE_JAVASCRIPT_MAP.get(from_extension, [])
+
+    for candidate in candidates:
+        first_extension = candidate["to_related"][0]
+        match = pathmap.find_paths(f"{base_path}{first_extension}", to_resources_index)
+        if not match:
+            continue
+
+        # Only create relations when the number of matches if inferior or equal to
+        # the current number of path segment matched.
+        if len(match.resource_ids) > match.matched_path_length:
+            continue
+
+        for resource_id in match.resource_ids:
+            first_match = to_resources.get(id=resource_id)
+
+            reversed_parent_segment = pathmap.get_reversed_path_segments(
+                str(Path(first_match.path).parent)
+            )
+
+            # Collect the transformed files within the same parent directory.
+            matches = []
+            for extension in candidate["to_related"]:
+                reversed_segments = [f"{basename}{extension}"] + reversed_parent_segment
+                resource_key = pathmap.convert_segments_to_path(reversed_segments)
+                if to_resources_index.exists(resource_key):
+                    value = to_resources_index.get(resource_key)
+                    matches.append(to_resources.get(id=value[1][0]))
+
+            is_compiled = js.is_minified_and_map_compiled_from_source(
+                matches,
+                from_resource,
+                minified_extension=candidate["to_minified_ext"],
+            )
+
+            if not is_compiled and bool(candidate["to_minified_ext"]):
+                continue
+
+            map_type = "js_compiled" if is_compiled else "path"
```

Comment:
I'm not sure its a good idea to use "path" here, as we do not have a was after to distinguish relations created in the JS context or in the map_path one.
What about using a different type, such as `js_path`?

Reply:
Why not using the same `js_compiled` by the way? Not sure about the value of using different map_type within the same mapper.

Reply:
`js_compiled` is used when we have found either the source or sourceContent in the map file corresponds to a file on `from` side. However, there are scenarios where we don't have any map file. For example, `.d.ts` files are related to `.ts` files, but we don't have any map file. In such cases, the only option is to compare the file paths and it would be incorrect to label those relationship as `js_compiled`.

Reply:
Thanks for the explanation. Let's use `js_path` as it's a bit more specialized than the generic `path`.
Also, a comment about this in the code around this line would be helpful.

Reply:
Done